kafka-client/src/main/java/dev/responsive/kafka/internal/db/partitioning/SubPartitioner.java

```diff
@@ -16,96 +16,126 @@
+ * and ending with {@code k-partition * n + n} (exclusive). To
  * compute which new partition a key falls into, it applies a hash
  * function and mods by {@code n}, adding that value to the base
- * mapped partition.</p>
+ * mapped partition.
  */
-public class SubPartitioner {
+public class SubPartitioner implements ResponsivePartitioner<Bytes, Integer> {
 
   public static final SubPartitioner NO_SUBPARTITIONS = new SubPartitioner(1, k -> 0);
 
   /**
-   * the number of subpartitions is: {@code original_partitions * n}
+   * the number of subpartitions is: {@code kafka_partitions * n}
    */
   private final int factor;
   private final Function<Bytes, Integer> hasher;
 
   public static SubPartitioner create(
-      final OptionalInt actualRemoteCount,
-      final int kafkaPartitions,
-      final int desiredNum,
-      final TableName name,
-      final String changelogTopicName,
-      final Hasher hasher
+      final int numKafkaPartitions,
+      final ResponsiveKeyValueParams params,
+      final ResponsiveConfig config,
+      final String changelogTopicName
   ) {
-    final int factor = (desiredNum == ResponsiveConfig.NO_SUBPARTITIONS)
-        ? 1 : (int) Math.ceil((double) desiredNum / kafkaPartitions);
-    final int computedRemoteNum = factor * kafkaPartitions;
-
-    if (actualRemoteCount.isPresent() && actualRemoteCount.getAsInt() != computedRemoteNum) {
-      throw new ConfigException(String.format("%s was configured to %d, which "
-              + "given %s partitions in kafka topic %s would result in %d remote partitions "
-              + "for table %s (remote partitions must be a multiple of the kafka partitions). "
-              + "The remote store is already initialized with %d partitions - it is backwards "
-              + "incompatible to change this. Please set %s to %d.",
-          ResponsiveConfig.STORAGE_DESIRED_NUM_PARTITION_CONFIG, desiredNum, kafkaPartitions,
-          changelogTopicName, computedRemoteNum, name.kafkaName(),
-          actualRemoteCount.getAsInt(), ResponsiveConfig.STORAGE_DESIRED_NUM_PARTITION_CONFIG,
-          actualRemoteCount.getAsInt()));
-    }
+    if (params.schemaType() == SchemaTypes.KVSchema.FACT) {
+      return NO_SUBPARTITIONS;
+    } else {
+      // TODO(agavra): write the actual remote partition count into cassandra
+      final OptionalInt actualRemoteCount = OptionalInt.empty();
+
+      final int requestedNumSubPartitions = config.getInt(STORAGE_DESIRED_NUM_PARTITION_CONFIG);
+
+      final int factor = (requestedNumSubPartitions == ResponsiveConfig.NO_SUBPARTITIONS)
+          ? 1 : (int) Math.ceil((double) requestedNumSubPartitions / numKafkaPartitions);
+      final int computedRemoteNum = factor * numKafkaPartitions;
 
-    return new SubPartitioner(factor, hasher);
+      if (actualRemoteCount.isPresent() && actualRemoteCount.getAsInt() != computedRemoteNum) {
+        throw new ConfigException(String.format("%s was configured to %d, which "
```

Comment:
I think checkstyle will catch this, but these lines seem a bit long :) (reviewing side-by-side on GH)

Reply:
indeed they were 😛